File: Prism/App/TranslationActionPolicy.swift
import Foundation

/// Resolved UI action after translation: panel vs in-place replace.
struct TranslationActionResolution: Equatable {
    let showPanel: Bool
    let canReplace: Bool
    let sendAfter: Bool
}

/// Maps capture context + hotkey presentation to panel vs replace-in-place.
///
/// Invariants:
/// - Language pair (`usedOutgoingPair`) never affects this policy — only capture + presentation.
/// - Replace vs panel follows `capture.isEditable` (AX + clipboard axError context at capture time).
/// - No mouse/window-band heuristics — cursor position is unpredictable for the user.
/// - Confirmed read-only with `.replaceInPlace` → panel, Copiar only.
enum TranslationActionPolicy {

    /// True when the translation should be pasted into the focused field instead of the panel.
    static func shouldReplaceInField(
        preferReplaceInPlace: Bool,
        capture: FocusedTextCapture
    ) -> Bool {
        capture.isEditable
    }

    static func resolve(
        presentation: TranslationPresentation,
        capture: FocusedTextCapture
    ) -> TranslationActionResolution {
        let preferReplaceInPlace: Bool
        switch presentation {
        case .replaceInPlace:
            preferReplaceInPlace = true
        case .popup:
            preferReplaceInPlace = false
        }

        let replaceInField = shouldReplaceInField(
            preferReplaceInPlace: preferReplaceInPlace,
            capture: capture
        )

        switch presentation {
        case .replaceInPlace(let sendAfter):
            return TranslationActionResolution(
                showPanel: !replaceInField,
                canReplace: false,
                sendAfter: sendAfter
            )
        case .popup:
            return TranslationActionResolution(
                showPanel: true,
                canReplace: replaceInField,
                sendAfter: false
            )
        }
    }
}

File: Prism.Windows/src/Prism.Core/Policy/TranslationActionPolicy.cs
using Prism.Core.TextIO;

namespace Prism.Core.Policy;

public enum TranslationPresentationKind
{
    ReplaceInPlace,
    Popup,
}

public sealed record TranslationPresentation(TranslationPresentationKind Kind, bool SendAfter = false)
{
    public static TranslationPresentation ReplaceInPlace(bool sendAfter = false) =>
        new(TranslationPresentationKind.ReplaceInPlace, sendAfter);

    public static TranslationPresentation Popup() =>
        new(TranslationPresentationKind.Popup);
}

public enum TranslationTriggerKind
{
    HotkeyTranslateOnly,
    HotkeyTranslateAndSend,
    HotkeyPopup,
    EnterKey,
}

public sealed record TranslationTrigger(TranslationTriggerKind Kind, string? Chord = null)
{
    public string Title => Kind switch
    {
        TranslationTriggerKind.HotkeyTranslateOnly => $"hotkey «Translate» ({Chord})",
        TranslationTriggerKind.HotkeyTranslateAndSend => $"hotkey «Translate and send» ({Chord})",
        TranslationTriggerKind.HotkeyPopup => $"hotkey «Panel» ({Chord})",
        TranslationTriggerKind.EnterKey => "Enter key (translate-and-send mode)",
        _ => Kind.ToString(),
    };

    public string ExpectedOutcome => Kind switch
    {
        TranslationTriggerKind.HotkeyTranslateOnly => "translate and replace in field, without sending",
        TranslationTriggerKind.HotkeyTranslateAndSend or TranslationTriggerKind.EnterKey =>
            "translate, replace, and send (Enter)",
        TranslationTriggerKind.HotkeyPopup => "show result in floating panel",
        _ => "",
    };
}

public sealed record TranslationActionResolution(bool ShowPanel, bool CanReplace, bool SendAfter);

/// <summary>
/// Maps capture context + hotkey presentation to panel vs replace-in-place.
/// Language pair never affects this policy — only capture + presentation.
/// </summary>
public static class TranslationActionPolicy
{
    public static bool ShouldReplaceInField(bool preferReplaceInPlace, FocusedTextCapture capture) =>
        capture.IsEditable;

    public static TranslationActionResolution Resolve(
        TranslationPresentation presentation,
        FocusedTextCapture capture)
    {
        var preferReplaceInPlace = presentation.Kind == TranslationPresentationKind.ReplaceInPlace;
        var replaceInField = ShouldReplaceInField(preferReplaceInPlace, capture);

        return presentation.Kind switch
        {
            TranslationPresentationKind.ReplaceInPlace => new TranslationActionResolution(
                ShowPanel: !replaceInField,
                CanReplace: false,
                SendAfter: presentation.SendAfter),
            TranslationPresentationKind.Popup => new TranslationActionResolution(
                ShowPanel: true,
                CanReplace: replaceInField,
                SendAfter: false),
            _ => new TranslationActionResolution(true, false, false),
        };
    }
}
